core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -448,45 +447,84 @@ private RexNode simplifyCase(RexCall call) {
           return disjunction;
         }
       }
-      // 2) Another simplification
-      //   CASE
-      //   WHEN p1 THEN TRUE
-      //   WHEN p2 THEN FALSE
-      //   WHEN p3 THEN TRUE
-      //   ELSE FALSE
-      //   END
-      // if p1...pn cannot be nullable
-      for (Ord<Pair<RexNode, RexNode>> pair : Ord.zip(pairs)) {
-        if (pair.e.getKey().getType().isNullable()) {
-          break trueFalse;
+      boolean anyNullable = false;
+    trueFalse:
```

Comment:
I think we can kill the trueFalse tag and make the code linear (or maybe I am not reading it properly with the indentation)? It would be more readable, I think after your changes in the method, it does not make sense to have it anymore. 1) Check nullability. 2) Try 2. 3) Try 3.